src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java

```diff
@@ -358,14 +361,13 @@ public void write(String text) {
 					if (this.connection.rendezvousWebSocket != null) {
 						flushReason = FlushReason.RENDEZVOUS_EXISTS;
 					} else {
-						int bufferedCount = this.writeBufferStream != null ? this.writeBufferStream.position() : 0;
+						int bufferedCount = this.writeBufferStream != null ? this.writeBufferStream.size() : 0;
 						if (count + bufferedCount <= MAX_CONTROL_CONNECTION_BODY_SIZE) {
 
 							// There's still a chance we might be able to respond over the control
 							// connection, accumulate bytes
 							if (this.writeBufferStream == null) {
-								int initialStreamSize = Math.min(count, MAX_CONTROL_CONNECTION_BODY_SIZE);
-								this.writeBufferStream = ByteBuffer.allocate(initialStreamSize);
+								this.writeBufferStream = new ByteArrayOutputStream(count);
```

Comment:
>new ByteArrayOutputStream(count) [](start = 33, length = 32)

ByteArrayOutputStream can auto increase capacity, unlike ByteBuffer. Used to cause exception here with ByteBuffer if multiple writes were called that were under 64kb combined.

Reply:
It appears that ByteArrayOutputStream is probably the correct thing to use since it will allocate a new array and copy all the existing data into the new array. It's better to let that built in class do it then to reinvent that wheel with our own code.

---
In reply to: [275564416](https://github.com/Azure/azure-relay-java/pull/39#discussion_r275564416) [](ancestors = 275564416)